File: dsl-api/src/main/java/cbs/dsl/api/HelperDefinition.java
package cbs.dsl.api;

import cbs.dsl.api.HelperTypes.HelperInput;
import cbs.dsl.api.HelperTypes.HelperOutput;

/**
 * Defines a helper — a reusable computation that can be invoked from event context blocks or
 * transactions.
 *
 * <p>Implementations are typically created via the Kotlin DSL {@code helper { }} block or annotated
 * with {@link DslComponent} for compile-time registration.
 */
public interface HelperDefinition extends StandardDslDefinition {

  /**
   * Executes this helper with the given typed input.
   *
   * @param input the helper input
   * @return the helper output
   */
  HelperOutput execute(HelperInput input);

  default HelperOutput preview(HelperInput input) {
    return execute(input);
  }
}

File: dsl-api/src/main/kotlin/cbs/dsl/api/HelperDefinition.kt
package cbs.dsl.api

interface HelperDefinition : HelperFunction<HelperInput, HelperOutput> {
  val code: String

  /**
   * Optional display name for this helper. Used to distinguish DSL overrides from the
   * underlying implementation class/bean identified by [code]. When set, the DSL block is
   * treated as a named override of the bean registered under [code].
   *
   * Example:
   * ```kotlin
   * helper("LOAN_CONDITIONS_BY_ID") {
   *     name("TestLoanConditions")
   *     execute { ctx -> mapOf("loanId" to ctx.params["loanId"], "currency" to "USD") }
   * }
   * ```
   */
  val name: String?
    get() = null

  /**
   * List of parameter definitions declared in the `parameters { }` block.
   * Used for validation and documentation purposes.
   */
  val parameters: List<ParameterDefinition>
    get() = emptyList()

  /**
   * Optional context enrichment block that runs before the execute block.
   * Allows helpers to enrich the context with additional data before execution.
   *
   * Example:
   * ```kotlin
   * helper("LOAN_HELPER") {
   *     context { ctx ->
   *         ctx["defaultCurrency"] = "USD"
   *     }
   *     execute { ctx -> mapOf("currency" to ctx["defaultCurrency"]) }
   * }
   * ```
   */
  val contextBlock: (cbs.dsl.api.context.HelperContext) -> Unit
    get() = {}
}
